plugins/builds/index.js

```diff
@@ -709,6 +764,67 @@ function getParentBuildIds({ currentBuildId, parentBuilds, joinListNames, pipeli
     return Array.from(new Set([currentBuildId, ...parentBuildIds]));
 }
 
+/**
+ * Delete nextBuild, create teardown build if it doesn't exist, and return teardown build or return null
+ * @param  {String}  nextJobName                  Next job name
+ * @param  {Object}  current                      Object with stage, event, pipeline info
+ * @param  {Object}  buildConfig                  Build config
+ * @param  {Factory} jobFactory                   Job factory
+ * @param  {Factory} buildFactory                 Build factory
+ * @param  {String}  username                     Username
+ * @param  {String}  scmContext                   Scm context
+ * @return {Array}                                Array of promises
+ */
+async function handleStageFailure({
+    nextJobName,
+    current,
+    buildConfig,
+    jobFactory,
+    buildFactory,
+    username,
+    scmContext
+}) {
+    const buildDeletePromises = [];
+    const stageTeardownName = getFullStageJobName({ stageName: current.stage.name, jobName: 'teardown' });
+
+    if (nextJobName === stageTeardownName) {
+        return buildDeletePromises;
+    }
+    if (buildConfig.eventId) {
+        // Remove next build
+        buildDeletePromises.push(deleteBuild(buildConfig, buildFactory));
+
+        // Check if stage teardown build already exists
+        const stageTeardownJob = await jobFactory.get({
```

Comment:
We should separate the concerns. Let us not interleave the responsibility of creating `teardown` build along side cleaning up of join builds.

Can we move the logic to create `teardown` build to separate function `checkAndCreateStageTeardownBuild()` and invoke this when a stage job fails?